site/content/coursework2.md

```diff
@@ -0,0 +1,424 @@
+`part1_euler_bernoulli.py` placed in the root of the repository.
+{{< /hint >}}
+
+
+The analytical solution is
+$$w(x) = \begin{cases}
+    \frac{x(4x^2-3)}{48k}, & \mbox{for } 0 < x < \frac{1}{2} \\
+    \frac{(x-1)(1-8x+4x^2)}{48k}, & \mbox{for } \frac{1}{2} < x < 1
+    \end{cases}$$
+but some constant is wrong in the above.
+
+{{< question "Part 1a questions" >}}
+1. How does the error in your numerical solution behave under grid
+   refinement? Can you explain what you see?
+{{< /question >}}
+
+
+
+
+## Part 2: The 3D Darcy Equation
+
+
+Discretise the equation
+$$
+\partial_t u - \nabla \cdot K(x, y, z) \nabla u = f(x, y, z)
+$$
+on the cubic domain $\Omega = [0, 1] \times [0, 1] \times [0, 1]$,
+using forward Euler as a timestepping scheme.
+
+To do this, create a class `Poisson7pt` that discretises the spatial
+operator using a 2nd order accurate 7-point stencil (as derived in lectures).
+$$
+-\nabla \cdot K(x, y, z) \nabla u.
+$$
+
+Using
+$$
+K(x, y, z) = cos(\pi y)
+$$
+$$
+f(x, y, z) =  (\pi+3\pi^2) \sin(\pi x)\sin(\pi y)\sin(\pi z),
+$$
+Dirichlet boundary conditions on the boundary
+$$
+u(x, y, z) = 0.
+$$
+The exact solution for this
+problem is
+$$
+u^*(x, y, z) = \sin(\pi x)\sin(\pi y)\sin(\pi z).
+$$
+
+
+### Part 2a: timestepping
+{{< hint info >}}
+You should do your implementation for this part in a file called
+`part2_explicit_euler.py` placed in the root of the repository.
+{{< /hint >}}
+
+Implement an explicit Euler time-stepping scheme using your 7-point stencil.
+
+{{< hint info >}}
+Ensure that your implementation works correctly in parallel as well as
+serial (when run with MPI).
+{{< /hint >}}
+
+{{< question "Part 2a questions" >}}
+1. How does the error in your numerical solution behave under grid
+   refinement? Can you explain what you see?
+2. What restriction, if any, is there on the size of the timestep you can
+   choose?
+{{< /question >}}
+
+### Part 2b: A higher-order scheme
+
+{{< hint info >}}
+This part doesn't need to you to write any code.
+{{< /hint >}}
+
+{{< question "Part 2b questions" >}}
+
+1. Derive, but do not implement, the 4th order accurate stencil for a
+   5-point discretisation of the Laplace operator in one dimension.
+1. How do you have to modify the stencil at the boundary to maintain the
+   accuracy for:
+    1. Dirichlet conditions $u = g$?
+    1. Neumann conditions $\nabla u \cdot n = h$?
+1. Would this spatial discretisation have the same timestep restriction
+   as the 2nd order operator, or a different one? Explain your answer.
+
+{{< /question >}}
+
+## Part 3: multigrid
+
+### Part 3a: completing the multigrid solver
+Implement the missing pieces in the `MGSolver` class, namely
+
+- `MGSolver.jacobi`
+- `MGSolver.vcycle`
+- `MGSolver.wcycle`
+
+You should do this directly in the `mgsolver/mgsolver.py` file (don't
+forget to commit it!). If you do this correctly, the one dimensional
+tests should now pass.
+
+{{< hint info >}}

```

Comment:
As above, I wonder if this should be step 1.

Reply:
I've asked here to do try this with the 1a component.